### DISCUSSION TITLE
What is integration_steps referring to?

### DISCUSSION THREAD

**Discussion by kx-au** [blackjax-devs/blackjax#134] [Q&A] [answered]
What does the chain output `info.integration_steps` refer to precisely? In the documentation it says the number of steps in the full trajectory = number of integration steps taken. Say the value is N, does that mean:
(1) The current sampler position is N steps away from the previous position? (e.g. a standard HMC can take N steps to achieve the same integration length)
(2) N is the total number of integration steps required for one NUTS iteration, where the final sampled position is <N steps away from the previous position, e.g. N//2? (similar to how the standard NUTS generate the tree but only really uses half of it for the so-called "final trajectory")

On that note, what would be the number of gradient evaluations for one NUTS iteration (in terms of N)?

Thank you so much!

**Comment by rlouf** [comment, marked as answer]
`integration_steps` (which really should be named `num_integration_steps`) corresponds to the number of times the symplectic integrator was run building the NUTS trajectory (including the "wasted" half). So for the [velocity verlet integrator](https://github.com/blackjax-devs/blackjax/blob/08ea201796b93a79f13ac3a175731b0c858a27a4/blackjax/inference/hmc/integrators.py#L32) (default) `N` would be the number of gradient evaluation.

Indeed you can see that the number returned [here](https://github.com/blackjax-devs/blackjax/blob/08ea201796b93a79f13ac3a175731b0c858a27a4/blackjax/nuts.py#L216) is the length of the full trajectory that was built.

Does that make sense? I'll open an issue to make sure this is better documented. Thank you!

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

Still accurate: `info.num_integration_steps` is the number of times the symplectic integrator ran while building the NUTS trajectory (including the rejected half), so for velocity-Verlet it equals the number of gradient evaluations. The integrator source has since moved to `blackjax/mcmc/integrators.py` (the old `blackjax/inference/hmc/...` path is pre-restructure).
